src/3dDiff.c

```diff
@@ -0,0 +1,345 @@
+    AFNI_logger("3dDiff", argc, argv);
+
+    /* Parse the args */
+    if (argc == 1) { help_3dDiff(); exit(0); } /* No args supplied */
+
+    while ( iarg < argc && argv[iarg][0] == '-'){
+        /* OPTIONAL: tolerance */
+        if ( strncmp(argv[iarg],"-tol",4) == 0) {
+            if (iarg >= argc) ERROR_exit("Need value after -tol");
+            if ( sscanf( argv[++iarg] , "%f", &tol ) != 1 ) {
+                ERROR_exit("Cannot parse tolerance!\n") ;
+            }
+            iarg++; continue;
+        }
+        
+        /* OPTIONAL: brutalist */
+        if ( strcmp(argv[iarg],"-brutalist") == 0) {
+            brutalist = 1;
+            iarg++; continue;
+        }
+
+        /* OPTIONAL: tabular */
+        if ( strcmp(argv[iarg],"-tabular") == 0) {
+            tabular = 1;
+            iarg++; continue;
+        }
+
+        /* OPTIONAL: quiet */
+        if ( strcmp(argv[iarg],"-q") == 0) {
+            quiet = 1;
+            iarg++; continue;
+        }
+
+        /* OPTIONAL: tps_report */
+        if ( strcmp(argv[iarg],"-tps_report") == 0) {
+            tps_report = 1;
+            iarg++; continue;
+        }
+
+        /* left image */
+        if ( strncmp(argv[iarg],"-a",2) == 0) {
+            if (++iarg >= argc) ERROR_exit("Need dset after -a");
+            a_fname = strdup(argv[iarg]);
+            iarg++; continue;
+        }
+
+        /* right image */
+        if ( strncmp(argv[iarg],"-b",2) == 0) {
+            if (++iarg >= argc) ERROR_exit("Need dset after -b");
+            b_fname = strdup(argv[iarg]);
+            iarg++; continue;
+        }
+
+        ERROR_message("ILLEGAL option: %s\n", argv[iarg]) ;
+                suggest_best_prog_option(argv[0], argv[iarg]);
+        exit(1);
+    }
+
+    /* Validate args */
+    if ( !a_fname )
+        ERROR_exit("No dset a supplied!");
+    if ( !b_fname )
+        ERROR_exit("No dset b supplied!");
+    /* tolerance is guaranteed to exist, validation performed above */
+
+    /*   Thanks to PT for simplifying options for me */
+    disp_opt_sum = brutalist + tabular + quiet + tps_report;
+    if ( !disp_opt_sum ){
+        report = 1;
+    }
+    else if ( disp_opt_sum > 1 ){
+       ERROR_exit("Must choose ONLY one of these display opts:"
+                  "  -brutalist, -tabular, -q, -tps_report");
+    }
+    // ... and in any other case, one should have exactly 1 option
+    // entered by the user, and be OK
+
+
+    /* Load the images and check for mutual compatibility*/
+    set_obliquity_report(0); /* We'll check that ourselves below */
+    ds1 = THD_open_dataset(a_fname);
+    if ( ds1 == NULL ) ERROR_exit("Can't open dataset %s", a_fname);
+    DSET_load(ds1); CHECK_LOAD_ERROR(ds1);
+    ds2 = THD_open_dataset(b_fname);
+    if ( ds2 == NULL ) ERROR_exit("Can't open dataset %s", b_fname);
+    DSET_load(ds2); CHECK_LOAD_ERROR(ds2);
+    if ( THD_dataset_mismatch( ds1, ds2 ) ) {
+        ERROR_exit("Mismatch between dsets!\n");
+    }
```

Comment:
I wonder if the text output here should differ based on the user's selected mode.  For example, if the user ran `3dDiff -1 ...`, then it would be nice to have a `-1` output here, rather than this failure?

For example, within this if condition of a mismatch, each different format for providing this information can be used, similar to output in a non-mismatch case:

```
if ( quiet ) { 
    printf("-1\n");
    exit(1);
}
if ( tabular ) {
    printf("-1 -1 -1 -1\n");
    exit(1);
}
if ( report || tps_report ) {
    ERROR_exit("Mismatch between dsets!\n");
}
```

This is useful because in quiet or brutalist modes, the user only expects a single number or 4 numbers output, respectively---so that should be given to them.

Reply:
That makes sense, I'll patch that shortly.

Reply:
In the most recent commit, the `diff3d_crash` macro handles this.